src/game/logic/object/weapon.cpp

```diff
@@ -954,6 +1520,214 @@ void WeaponTemplate::Deal_Damage_Internal(ObjectID source_id,
     const WeaponBonus &bonus,
     bool is_projectile_detonation) const
 {
-}
+    if (source_id != OBJECT_UNK && (victim_id != OBJECT_UNK || pos != nullptr)) {
+        Object *source = g_theGameLogic->Find_Object_By_ID(source_id);
+        Trim_Old_Historic_Damage();
+
+        if (m_historicBonusCount > 0 && m_historicBonusWeapon != this) {
+            float radius_sqr = m_historicBonusRadius * m_historicBonusRadius;
+            int count = 0;
+            unsigned int frame = g_theGameLogic->Get_Frame();
+            unsigned int frame2 = frame - m_historicBonusTime;
+
+            for (auto it = m_historicDamage.begin(); it != m_historicDamage.end(); it++) {
+                if (it->frame >= frame2) {
+                    if (Is_2D_Dist_Squared_Less_Than(pos, &it->location, radius_sqr)) {
+                        count++;
+                    }
+                }
+            }
 
-#endif
+            if (count < m_historicBonusCount - 1) {
+                HistoricWeaponDamageInfo info(frame, *pos);
+                m_historicDamage.push_back(info);
+            } else {
+                g_theWeaponStore->Create_And_Fire_Temp_Weapon(m_historicBonusWeapon, source, pos);
+                m_historicDamage.clear();
+            }
+        }
+
+        Object *victim;
+
+        if (victim_id != OBJECT_UNK) {
+            victim = g_theGameLogic->Find_Object_By_ID(victim_id);
+        } else {
+            victim = nullptr;
+        }
+
+        if (victim != nullptr) {
+            pos = victim->Get_Position();
+        }
+
+        DamageType damage = Get_Damage_Type();
+        DeathType death = Get_Death_Type();
+        ObjectStatusTypes status = Get_Damage_Status_Type();
+
+        if (Get_Projectile_Template() == nullptr || is_projectile_detonation) {
+            float primary_radius = Get_Primary_Damage_Radius(bonus);
+            float secondary_radius = Get_Secondary_Damage_Radius(bonus);
+            float primary_damage = Get_Primary_Damage(bonus);
+            float secondary_damage = Get_Secondary_Damage(bonus);
+            int mask = Get_Affects_Mask();
+
+            captainslog_dbgassert(secondary_radius > primary_radius || secondary_radius == 0.0f,
```

Comment:
Assert condition does not match assert message.

Reply:
fixed.